File: apps/cosmos/internal/ocpshuffle/shuffle_v1_test.go
package ocpshuffle

import (
	"testing"

	"onchainpoker/apps/cosmos/internal/ocpcrypto"
)

func makeDeck(pk ocpcrypto.Point, n int, seed uint64) []ocpcrypto.ElGamalCiphertext {
	deck := make([]ocpcrypto.ElGamalCiphertext, 0, n)
	for i := 0; i < n; i++ {
		m := ocpcrypto.MulBase(ocpcrypto.ScalarFromUint64(uint64(i + 1)))
		r := ocpcrypto.ScalarFromUint64(seed + uint64(i+1))
		ct, err := ocpcrypto.ElGamalEncrypt(pk, m, r)
		if err != nil {
			panic(err)
		}
		deck = append(deck, ct)
	}
	return deck
}

func TestShuffleV1_ValidProofVerifiesSmallDeck(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(42)
	pk := ocpcrypto.MulBase(sk)
	deckIn := makeDeck(pk, 10, 123)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 7
	}
	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: 10})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}
	vr := ShuffleVerifyV1(pk, deckIn, res.ProofBytes, nil)
	if !vr.OK {
		t.Fatalf("verify failed: %s", vr.Error)
	}
	if len(vr.DeckOut) != 10 {
		t.Fatalf("deckOut length mismatch: %d", len(vr.DeckOut))
	}
}

func TestShuffleV1_VerifySupportsOddAndEvenDeckSizes(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(2468)
	pk := ocpcrypto.MulBase(sk)

	for _, n := range []int{2, 3, 4, 5, 6} {
		deckIn := makeDeck(pk, n, 1000+uint64(n))
		seed := make([]byte, 32)
		for i := range seed {
			seed[i] = byte(17 + n)
		}

		res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: 7})
		if err != nil {
			t.Fatalf("prove n=%d: %v", n, err)
		}
		vr := ShuffleVerifyV1(pk, deckIn, res.ProofBytes, nil)
		if !vr.OK {
			t.Fatalf("verify n=%d failed: %s", n, vr.Error)
		}
		if len(vr.DeckOut) != n {
			t.Fatalf("deckOut length mismatch n=%d: %d", n, len(vr.DeckOut))
		}
	}
}

func TestShuffleV1_TamperingOutputDeckBytesFailsVerification(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(123)
	pk := ocpcrypto.MulBase(sk)
	deckIn := makeDeck(pk, 12, 999)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 9
	}
	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: 12})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}

	bad := make([]byte, len(res.ProofBytes))
	copy(bad, res.ProofBytes)
	// Header is 1 + 2 + 2 = 5 bytes; deck snapshot begins immediately.
	bad[5+0] ^= 0x01

	vr := ShuffleVerifyV1(pk, deckIn, bad, nil)
	if vr.OK {
		t.Fatalf("expected verify to fail")
	}
}

func TestShuffleV1_WrongPermutationSwappingTwoCiphertextsFailsVerification(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(321)
	pk := ocpcrypto.MulBase(sk)
	n := 10
	deckIn := makeDeck(pk, n, 222)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 8
	}
	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: n})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}

	bad := make([]byte, len(res.ProofBytes))
	copy(bad, res.ProofBytes)
	headerLen := 5
	ctLen := 64
	a0 := append([]byte(nil), bad[headerLen+0*ctLen:headerLen+1*ctLen]...)
	a1 := append([]byte(nil), bad[headerLen+1*ctLen:headerLen+2*ctLen]...)
	copy(bad[headerLen+0*ctLen:], a1)
	copy(bad[headerLen+1*ctLen:], a0)

	vr := ShuffleVerifyV1(pk, deckIn, bad, nil)
	if vr.OK {
		t.Fatalf("expected verify to fail")
	}
}

func TestShuffleV1_MissingRerandomizationReusingC1FailsVerification(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(777)
	pk := ocpcrypto.MulBase(sk)
	deckIn := makeDeck(pk, 8, 111)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 3
	}
	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: 8})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}

	bad := make([]byte, len(res.ProofBytes))
	copy(bad, res.ProofBytes)

	headerLen := 5
	ctLen := 64
	in0c1 := deckIn[0].C1.Bytes()
	copy(bad[headerLen+0*ctLen:], in0c1)

	vr := ShuffleVerifyV1(pk, deckIn, bad, nil)
	if vr.OK {
		t.Fatalf("expected verify to fail")
	}
}

func TestShuffleV1_N52SmokeRounds10Verifies(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(999)
	pk := ocpcrypto.MulBase(sk)
	deckIn := makeDeck(pk, 52, 555)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 1
	}
	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: 10})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}

	vr := ShuffleVerifyV1(pk, deckIn, res.ProofBytes, nil)
	if !vr.OK {
		t.Fatalf("verify failed: %s", vr.Error)
	}
}

// ---- WS5 ctx-binding (v2) tests ----

func mustCtx(t *testing.T, tableID uint64, handID uint64, round uint16, shuffler string) []byte {
	t.Helper()
	ctx, err := BuildShuffleContext(tableID, handID, round, shuffler)
	if err != nil {
		t.Fatalf("BuildShuffleContext: %v", err)
	}
	return ctx
}

func TestShuffleV1_CtxBindingV2VerifiesUnderMatchingContext(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(4242)
	pk := ocpcrypto.MulBase(sk)
	n := 6
	deckIn := makeDeck(pk, n, 7777)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 11
	}
	ctx := mustCtx(t, 1, 2, 3, "cosmosvaloper1foo")

	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: n, Context: ctx})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}
	// Header byte 0 must be version=2 (v2 format).
	if res.ProofBytes[0] != ShuffleProofV2Version {
		t.Fatalf("expected v2 header, got %d", res.ProofBytes[0])
	}

	vr := ShuffleVerifyV1(pk, deckIn, res.ProofBytes, ctx)
	if !vr.OK {
		t.Fatalf("verify failed: %s", vr.Error)
	}
}

func TestShuffleV1_CtxBindingProofBoundToCtxADoesNotVerifyUnderCtxB(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(4242)
	pk := ocpcrypto.MulBase(sk)
	n := 6
	deckIn := makeDeck(pk, n, 7777)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 12
	}
	ctxA := mustCtx(t, 1, 2, 3, "cosmosvaloper1foo")
	ctxB := mustCtx(t, 1, 2, 3, "cosmosvaloper1bar") // different shuffler

	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: n, Context: ctxA})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}

	vr := ShuffleVerifyV1(pk, deckIn, res.ProofBytes, ctxB)
	if vr.OK {
		t.Fatalf("expected verify to fail under ctxB")
	}
}

func TestShuffleV1_CtxBindingDifferentHandIDRejected(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(4242)
	pk := ocpcrypto.MulBase(sk)
	n := 6
	deckIn := makeDeck(pk, n, 7777)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 13
	}
	ctx2 := mustCtx(t, 1, 2, 1, "v")
	ctx3 := mustCtx(t, 1, 3, 1, "v") // different handId

	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: n, Context: ctx2})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}

	vr := ShuffleVerifyV1(pk, deckIn, res.ProofBytes, ctx3)
	if vr.OK {
		t.Fatalf("expected verify to fail under ctx3")
	}
}

func TestShuffleV1_CtxBindingEmptyContextRejectedAtV2(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(4242)
	pk := ocpcrypto.MulBase(sk)
	deckIn := makeDeck(pk, 4, 7777)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 14
	}
	_, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: 4, Context: []byte{}})
	if err == nil {
		t.Fatalf("expected empty context to be rejected")
	}
}

func TestShuffleV1_CtxBindingV2WithoutCallerCtxFailsVerify(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(4242)
	pk := ocpcrypto.MulBase(sk)
	n := 4
	deckIn := makeDeck(pk, n, 7777)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 15
	}
	ctx := mustCtx(t, 1, 1, 1, "v")
	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: n, Context: ctx})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}

	vr := ShuffleVerifyV1(pk, deckIn, res.ProofBytes, nil)
	if vr.OK {
		t.Fatalf("expected verify to fail when verifier has no context")
	}
}

func TestShuffleV1_CtxBindingV1ProofBackwardCompatible(t *testing.T) {
	// Emits v1 (no context), verifies with nil.
	sk := ocpcrypto.ScalarFromUint64(42)
	pk := ocpcrypto.MulBase(sk)
	deckIn := makeDeck(pk, 10, 123)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 7
	}
	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: 10})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}
	if res.ProofBytes[0] != ShuffleProofV1Version {
		t.Fatalf("expected v1 header, got %d", res.ProofBytes[0])
	}
	vr := ShuffleVerifyV1(pk, deckIn, res.ProofBytes, nil)
	if !vr.OK {
		t.Fatalf("verify failed: %s", vr.Error)
	}
}

func TestShuffleV1_CtxBindingV1ProofRejectedWhenCtxSupplied(t *testing.T) {
	sk := ocpcrypto.ScalarFromUint64(42)
	pk := ocpcrypto.MulBase(sk)
	deckIn := makeDeck(pk, 10, 123)

	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 7
	}
	res, err := ShuffleProveV1(pk, deckIn, ShuffleProveOpts{Seed: seed, Rounds: 10})
	if err != nil {
		t.Fatalf("prove: %v", err)
	}

	ctx := mustCtx(t, 1, 1, 1, "v")
	vr := ShuffleVerifyV1(pk, deckIn, res.ProofBytes, ctx)
	if vr.OK {
		t.Fatalf("expected verify to fail when v1 proof is given a context")
	}
}

func TestShuffleV1_BuildShuffleContextWireFormat(t *testing.T) {
	// Canonical wire format example (documented in shuffle_v1.go):
	//   u64le(tableId=1) || u64le(handId=2) || u16le(round=3) ||
	//   u16le(shufflerLen=17) || "cosmosvaloper1foo"
	got, err := BuildShuffleContext(1, 2, 3, "cosmosvaloper1foo")
	if err != nil {
		t.Fatalf("BuildShuffleContext: %v", err)
	}
	want := []byte{
		1, 0, 0, 0, 0, 0, 0, 0, // tableId=1 LE
		2, 0, 0, 0, 0, 0, 0, 0, // handId=2 LE
		3, 0, // round=3 LE
		17, 0, // shufflerLen=17 LE
	}
	want = append(want, []byte("cosmosvaloper1foo")...)
	if len(got) != len(want) {
		t.Fatalf("ctx length mismatch: got %d want %d", len(got), len(want))
	}
	for i := range want {
		if got[i] != want[i] {
			t.Fatalf("ctx byte %d mismatch: got %x want %x", i, got[i], want[i])
		}
	}
}

func TestShuffleV1_CtxBindingMatchesTSWireFormat(t *testing.T) {
	// Parity check: a specific (tableId, handId, round, shuffler) must hash
	// to the same bytes as the TS side. Length = 8+8+2+2+len(shuffler).
	ctx, err := BuildShuffleContext(0x0102030405060708, 0x0a0b0c0d0e0f1011, 0x1213, "abc")
	if err != nil {
		t.Fatalf("BuildShuffleContext: %v", err)
	}
	if len(ctx) != 8+8+2+2+3 {
		t.Fatalf("unexpected ctx length: %d", len(ctx))
	}
	// first 8 bytes = tableId LE
	expected0 := []byte{0x08, 0x07, 0x06, 0x05, 0x04, 0x03, 0x02, 0x01}
	for i, b := range expected0 {
		if ctx[i] != b {
			t.Fatalf("tableId LE byte %d: got %x want %x", i, ctx[i], b)
		}
	}
	// next 8 bytes = handId LE
	expected1 := []byte{0x11, 0x10, 0x0f, 0x0e, 0x0d, 0x0c, 0x0b, 0x0a}
	for i, b := range expected1 {
		if ctx[8+i] != b {
			t.Fatalf("handId LE byte %d: got %x want %x", i, ctx[8+i], b)
		}
	}
	// round LE
	if ctx[16] != 0x13 || ctx[17] != 0x12 {
		t.Fatalf("round bytes mismatch: got %x %x", ctx[16], ctx[17])
	}
	// shufflerLen=3 LE
	if ctx[18] != 3 || ctx[19] != 0 {
		t.Fatalf("shufflerLen bytes mismatch")
	}
	if string(ctx[20:]) != "abc" {
		t.Fatalf("shuffler bytes mismatch")
	}
}

File: packages/ocp-shuffle/test/shuffle_v1.test.ts
import test from "node:test";
import assert from "node:assert/strict";

import { elgamalEncrypt, mulBase } from "@onchainpoker/ocp-crypto";
import { shuffleProveV1, shuffleVerifyV1, buildShuffleContext } from "../index.js";

function makeDeck(pk: any, n: number, seed: bigint): any[] {
  const deck: any[] = [];
  for (let i = 0; i < n; i++) {
    const m = mulBase(BigInt(i + 1));
    // Deterministic-ish randomness for encrypting test deck.
    const r = seed + BigInt(i + 1);
    deck.push(elgamalEncrypt(pk, m, r));
  }
  return deck;
}

function mutateSingleProofForRound(
  proofBytes: Uint8Array,
  n: number,
  rounds: number,
  targetRound: number,
  singleSlot: number,
): Uint8Array {
  const proof = new Uint8Array(proofBytes);
  const switchProofLen = 4 * 3 * 32;
  const singleProofLen = 3 * 32;
  let offset = 5;

  for (let round = 0; round < rounds; round++) {
    const start = round & 1;
    offset += n * 64;
    const pairCount = Math.floor((n - start) / 2);
    offset += pairCount * switchProofLen;

    if (round === targetRound) {
      if (n % 2 === 1) {
        if (singleSlot !== 0) {
          throw new Error(`invalid singleSlot=${singleSlot} for odd deck`);
        }
        proof[offset] ^= 0x01;
        return proof;
      }
      if (start === 1) {
        const slotOffset = singleSlot * singleProofLen;
        proof[offset + slotOffset] ^= 0x01;
        return proof;
      }
      throw new Error(`no singles at round=${round} for n=${n}, start=${start}`);
    }

    if (n % 2 === 1 || start === 1) {
      const singleCount = n % 2 === 1 ? 1 : 2;
      offset += singleCount * singleProofLen;
    }
  }

  throw new Error(`targetRound ${targetRound} out of range`);
}

test("WS5: valid shuffle proof verifies (small deck)", () => {
  const sk = 42n;
  const pk = mulBase(sk);
  const deckIn = makeDeck(pk, 10, 123n);

  const { proofBytes } = shuffleProveV1(pk, deckIn, { seedUnsafeForTestsOnly: new Uint8Array(32).fill(7), rounds: 10 });
  const vr = shuffleVerifyV1(pk, deckIn, proofBytes);
  if (!vr.ok) throw new Error(vr.error);
  assert.equal(vr.deckOut.length, 10);
});

test("WS5: tampering output deck bytes fails verification", () => {
  const sk = 123n;
  const pk = mulBase(sk);
  const deckIn = makeDeck(pk, 12, 999n);

  const { proofBytes } = shuffleProveV1(pk, deckIn, { seedUnsafeForTestsOnly: new Uint8Array(32).fill(9), rounds: 12 });

  // Flip one bit in the first post-round deck snapshot.
  const bad = new Uint8Array(proofBytes);
  // Header is 1 + 2 + 2 = 5 bytes; deck snapshot begins immediately.
  bad[5 + 0] ^= 0x01;

  const vr = shuffleVerifyV1(pk, deckIn, bad);
  assert.equal(vr.ok, false);
});

test("WS5: wrong permutation (swapping two ciphertexts) fails verification", () => {
  const sk = 321n;
  const pk = mulBase(sk);
  const n = 10;
  const deckIn = makeDeck(pk, n, 222n);

  const { proofBytes } = shuffleProveV1(pk, deckIn, { seedUnsafeForTestsOnly: new Uint8Array(32).fill(8), rounds: n });

  const bad = new Uint8Array(proofBytes);
  const headerLen = 5;
  const ctLen = 64;

  const a0 = bad.slice(headerLen + 0 * ctLen, headerLen + 1 * ctLen);
  const a1 = bad.slice(headerLen + 1 * ctLen, headerLen + 2 * ctLen);
  bad.set(a1, headerLen + 0 * ctLen);
  bad.set(a0, headerLen + 1 * ctLen);

  const vr = shuffleVerifyV1(pk, deckIn, bad);
  assert.equal(vr.ok, false);
});

test("WS5: missing rerandomization (reusing c1) fails verification", () => {
  const sk = 777n;
  const pk = mulBase(sk);
  const deckIn = makeDeck(pk, 8, 111n);

  const { proofBytes } = shuffleProveV1(pk, deckIn, { seedUnsafeForTestsOnly: new Uint8Array(32).fill(3), rounds: 8 });

  // Force output[0].c1 bytes to equal input[0].c1 bytes in round 0 deck snapshot.
  const bad = new Uint8Array(proofBytes);
  const headerLen = 5;
  const ctLen = 64;
  const in0c1 = deckIn[0].c1.toBytes();
  bad.set(in0c1, headerLen + 0 * ctLen + 0 /* c1 offset */);

  const vr = shuffleVerifyV1(pk, deckIn, bad);
  assert.equal(vr.ok, false);
});

test("WS5: verify handles odd/even rounds with mixed parity singles", () => {
  const sk = 2468n;
  const pk = mulBase(sk);
  const rounds = 7;

  const cases = [2, 3, 4, 5, 6];
  for (const n of cases) {
    const deckIn = makeDeck(pk, n, BigInt(1000 + n));
    const seed = new Uint8Array(32).fill(17 + n);
    const { proofBytes } = shuffleProveV1(pk, deckIn, { seedUnsafeForTestsOnly: seed, rounds });
    const vr = shuffleVerifyV1(pk, deckIn, proofBytes);
    if (!vr.ok) throw new Error(vr.error);
    assert.equal(vr.deckOut.length, n);
  }
});

test("WS5: N=52 smoke (rounds=10) verifies", () => {
  const sk = 999n;
  const pk = mulBase(sk);
  const deckIn = makeDeck(pk, 52, 555n);

  const { proofBytes } = shuffleProveV1(pk, deckIn, { seedUnsafeForTestsOnly: new Uint8Array(32).fill(1), rounds: 10 });
  const vr = shuffleVerifyV1(pk, deckIn, proofBytes);
  if (!vr.ok) throw new Error(vr.error);
});

test("WS5: even deck odd-start round uses two single proofs (round 1, n=2, rounds=2)", () => {
  const sk = 2024n;
  const pk = mulBase(sk);
  const deckIn = makeDeck(pk, 2, 2222n);

  const { proofBytes } = shuffleProveV1(pk, deckIn, { seedUnsafeForTestsOnly: new Uint8Array(32).fill(4), rounds: 2 });
  const vr = shuffleVerifyV1(pk, deckIn, proofBytes);
  if (!vr.ok) throw new Error(vr.error);
});

test("WS5: even deck odd-start rejects tampered first single proof (round 1 slot 0)", () => {
  const sk = 2024n;
  const pk = mulBase(sk);
  const deckIn = makeDeck(pk, 2, 2222n);

  const { proofBytes } = shuffleProveV1(pk, deckIn, { seedUnsafeForTestsOnly: new Uint8Array(32).fill(4), rounds: 2 });
  const bad = mutateSingleProofForRound(proofBytes, 2, 2, 1, 0);
  const vr = shuffleVerifyV1(pk, deckIn, bad);
  assert.equal(vr.ok, false);
});

test("WS5: even deck odd-start rejects tampered second single proof (round 1 slot 1)", () => {
  const sk = 2024n;
  const pk = mulBase(sk);
  const deckIn = makeDeck(pk, 2, 2222n);

  const { proofBytes } = shuffleProveV1(pk, deckIn, { seedUnsafeForTestsOnly: new Uint8Array(32).fill(4), rounds: 2 });
  const bad = mutateSingleProofForRound(proofBytes, 2, 2, 1, 1);
  const vr = shuffleVerifyV1(pk, deckIn, bad);
  assert.equal(vr.ok, false);
});

test("WS5 ctx-binding: v2 proof verifies under matching context", () => {
  const sk = 4242n;
  const pk = mulBase(sk);
  const n = 6;
  const deckIn = makeDeck(pk, n, 7777n);

  const ctx = buildShuffleContext({
    tableId: 1n,
    handId: 2n,
    round: 3,
    shuffler: "cosmosvaloper1foo",
  });

  const { proofBytes } = shuffleProveV1(pk, deckIn, {
    seedUnsafeForTestsOnly: new Uint8Array(32).fill(11),
    rounds: n,
    context: ctx,
  });

  // Header byte 0 must now be version=2 (v2 format).
  assert.equal(proofBytes[0], 2);

  const vr = shuffleVerifyV1(pk, deckIn, proofBytes, ctx);
  if (!vr.ok) throw new Error(vr.error);
  assert.equal(vr.deckOut.length, n);
});

test("WS5 ctx-binding: proof with context A does not verify under context B", () => {
  const sk = 4242n;
  const pk = mulBase(sk);
  const n = 6;
  const deckIn = makeDeck(pk, n, 7777n);

  const ctxA = buildShuffleContext({
    tableId: 1n,
    handId: 2n,
    round: 3,
    shuffler: "cosmosvaloper1foo",
  });
  const ctxB = buildShuffleContext({
    tableId: 1n,
    handId: 2n,
    round: 3,
    shuffler: "cosmosvaloper1bar", // different shuffler
  });

  const { proofBytes } = shuffleProveV1(pk, deckIn, {
    seedUnsafeForTestsOnly: new Uint8Array(32).fill(12),
    rounds: n,
    context: ctxA,
  });

  const vr = shuffleVerifyV1(pk, deckIn, proofBytes, ctxB);
  assert.equal(vr.ok, false);
});

test("WS5 ctx-binding: proof bound to (handId=2) rejected under (handId=3)", () => {
  const sk = 4242n;
  const pk = mulBase(sk);
  const n = 6;
  const deckIn = makeDeck(pk, n, 7777n);

  const ctx2 = buildShuffleContext({ tableId: 1n, handId: 2n, round: 1, shuffler: "v" });
  const ctx3 = buildShuffleContext({ tableId: 1n, handId: 3n, round: 1, shuffler: "v" });

  const { proofBytes } = shuffleProveV1(pk, deckIn, {
    seedUnsafeForTestsOnly: new Uint8Array(32).fill(13),
    rounds: n,
    context: ctx2,
  });

  const vr = shuffleVerifyV1(pk, deckIn, proofBytes, ctx3);
  assert.equal(vr.ok, false);
});

test("WS5 ctx-binding: empty context rejected at v2 (prover throws)", () => {
  const sk = 4242n;
  const pk = mulBase(sk);
  const deckIn = makeDeck(pk, 4, 7777n);

  assert.throws(() =>
    shuffleProveV1(pk, deckIn, {
      seedUnsafeForTestsOnly: new Uint8Array(32).fill(14),
      rounds: 4,
      context: new Uint8Array(0),
    }),
  );
});

test("WS5 ctx-binding: v2 proof without caller-supplied context fails verify", () => {
  const sk = 4242n;
  const pk = mulBase(sk);
  const n = 4;
  const deckIn = makeDeck(pk, n, 7777n);

  const ctx = buildShuffleContext({ tableId: 1n, handId: 1n, round: 1, shuffler: "v" });
  const { proofBytes } = shuffleProveV1(pk, deckIn, {
    seedUnsafeForTestsOnly: new Uint8Array(32).fill(15),
    rounds: n,
    context: ctx,
  });

  const vr = shuffleVerifyV1(pk, deckIn, proofBytes /* no context */);
  assert.equal(vr.ok, false);
});

test("WS5 ctx-binding: v1 proof (no context) still verifies via legacy path", () => {
  // Ensures backward-compat: legacy proofs emitted with no context opt
  // continue to work when verifier is called without a context.
  const sk = 42n;
  const pk = mulBase(sk);
  const deckIn = makeDeck(pk, 10, 123n);

  const { proofBytes } = shuffleProveV1(pk, deckIn, {
    seedUnsafeForTestsOnly: new Uint8Array(32).fill(7),
    rounds: 10,
  });
  // Version byte must remain 1 (no context supplied).
  assert.equal(proofBytes[0], 1);

  const vr = shuffleVerifyV1(pk, deckIn, proofBytes);
  if (!vr.ok) throw new Error(vr.error);
});

test("WS5 ctx-binding: v1 proof rejected when verifier is given a context", () => {
  const sk = 42n;
  const pk = mulBase(sk);
  const deckIn = makeDeck(pk, 10, 123n);

  const { proofBytes } = shuffleProveV1(pk, deckIn, {
    seedUnsafeForTestsOnly: new Uint8Array(32).fill(7),
    rounds: 10,
  });

  const ctx = buildShuffleContext({ tableId: 1n, handId: 1n, round: 1, shuffler: "v" });
  const vr = shuffleVerifyV1(pk, deckIn, proofBytes, ctx);
  assert.equal(vr.ok, false);
});

test("WS5 ctx-binding: buildShuffleContext produces canonical wire format", () => {
  // u64le(tableId=1) || u64le(handId=2) || u16le(round=3) || u16le(len) || shuffler
  const ctx = buildShuffleContext({
    tableId: 1n,
    handId: 2n,
    round: 3,
    shuffler: "cosmosvaloper1foo",
  });
  // 8 + 8 + 2 + 2 + 17 = 37
  assert.equal(ctx.length, 37);
  // tableId little-endian
  assert.deepEqual(Array.from(ctx.subarray(0, 8)), [1, 0, 0, 0, 0, 0, 0, 0]);
  // handId
  assert.deepEqual(Array.from(ctx.subarray(8, 16)), [2, 0, 0, 0, 0, 0, 0, 0]);
  // round=3
  assert.deepEqual(Array.from(ctx.subarray(16, 18)), [3, 0]);
  // shufflerLen=17
  assert.deepEqual(Array.from(ctx.subarray(18, 20)), [17, 0]);
  // utf-8 bytes
  assert.equal(new TextDecoder().decode(ctx.subarray(20)), "cosmosvaloper1foo");
});
